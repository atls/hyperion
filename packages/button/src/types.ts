export interface StyledButtonProps {
  borderRadius?: string | string[]
  height?: string | string[]
  width?: string | string[]
  padding?: string | string[]
  background?: string | string[]
  justifyContent?: string | string[]
  alignItems?: string | string[]
}
