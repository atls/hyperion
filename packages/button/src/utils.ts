import { Children, ReactNode, createElement } from 'react'

export const combinePureChildren = (children: ReactNode): ReactNode[] => {
  let isPrevChildPure: boolean = false
  const childList: ReactNode[] = []

  Children.forEach(children, child => {
    const isCurrentChildPure = typeof child === 'string' || typeof child === 'number'

    if (isPrevChildPure && isCurrentChildPure) {
      const lastIndex = childList.length - 1
      const lastChild = childList[lastIndex]
      childList[lastIndex] = [lastChild, child].join(' ')
    } else {
      childList.push(child)
    }

    isPrevChildPure = isCurrentChildPure
  })

  return childList
}

export const divideChildren = (
  children: ReactNode,
  Wrapper,
  props = {}
): ReactNode | ReactNode[] => {
  if (Children.count(children) > 1) {
    const combinedChildren = combinePureChildren(children)
    const count = combinedChildren.length

    return Children.map(combinedChildren, (child, index) => {
      if (index < count - 1) {
        return createElement(Wrapper, props, child)
      }

      return child
    })
  }

  return children
}
