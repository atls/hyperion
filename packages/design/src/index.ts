export * from './Sample'
export * from './Layout'
export * from './Container'
export * from './List'
export * from './Text'
export * from './Space'
