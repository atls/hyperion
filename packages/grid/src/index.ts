export * from './Column'
export * from './Grid'
