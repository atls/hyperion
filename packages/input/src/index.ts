export * from './Textarea'
export * from './Input'
export * from './Checkbox'
