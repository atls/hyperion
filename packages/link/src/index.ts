export * from './Link'
export * from './NextLink'
export * from './RouteLink'
