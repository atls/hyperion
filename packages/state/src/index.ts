export * from './withState'
export * from './useActive'
export * from './useHover'
export * from './useFocus'
