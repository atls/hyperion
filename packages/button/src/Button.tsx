import React                                       from 'react'
import styled                                      from '@emotion/styled'
import { border, color, flexbox, layout, padding } from 'styled-system'

import { StyledButtonProps }                       from './types'
import { divideChildren }                          from './utils'

const base: any = () => ({
  boxSizing: 'border-box',
  display: 'flex',
  whiteSpace: 'nowrap',
  cursor: 'pointer',
  border: 'none',
  outline: 'none',
  padding: 0,
  fontSize: 'inherit',
  fontWeight: 'inherit',
  lineHeight: 'inherit',
  fontFamily: 'inherit',
  textAlign: 'inherit',
  letterSpacing: 'inherit',
  color: 'inherit',
  '&:active': {
    transform: 'scale(0.98)',
  },
})

const StyledButton = styled('button')<StyledButtonProps>(
  base,
  border,
  flexbox,
  layout,
  padding,
  color
)

const ContentPart = styled.span({
  marginRight: '12px',
  display: 'flex',
})

const Button = ({ children, ...props }) => (
  <StyledButton {...props}>{divideChildren(children, ContentPart)}</StyledButton>
)

Button.defaultProps = {
  borderRadius: 'normal',
  justifyContent: 'center',
  alignItems: 'center',
}

export { Button }
