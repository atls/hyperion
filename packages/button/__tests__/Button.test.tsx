import React      from 'react'
import renderer   from 'react-test-renderer'

import { Button } from '../src/index'

it('Button should render correctly', () => {
  const defaultTheme = {
    colors: {
      green: '#0f0',
      black: '#000',
      white: '#fff',
      blue: '#00f',
      red: '#f00',
      lightGray: 'gray',
    },
    shadows: {
      controls: '0 0 10px rgba(0,0,0,0.5)',
    },
    borders: {
      controls: '2px solid black',
    },
  }

  const tree = renderer
    .create(
      <Button
        theme={defaultTheme}
        backgroundColor='white'
        hoverBackgroundColor='red'
        activeBackgroundColor='green'
      >
        Button
      </Button>,
    )
    .toJSON()

  expect(tree).toMatchSnapshot()
})
