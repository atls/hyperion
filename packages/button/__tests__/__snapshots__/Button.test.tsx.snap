// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button should render correctly 1`] = `
.emotion-0 {
  box-sizing: border-box;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  white-space: nowrap;
  cursor: pointer;
  border: none;
  outline: none;
  padding: 0;
  font-size: inherit;
  font-weight: inherit;
  line-height: inherit;
  font-family: inherit;
  text-align: inherit;
  -webkit-letter-spacing: inherit;
  -moz-letter-spacing: inherit;
  -ms-letter-spacing: inherit;
  letter-spacing: inherit;
  color: inherit;
  border-radius: normal;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  background-color: #fff;
}

.emotion-0:hover {
  background-color: #f00;
}

.emotion-0:active {
  -webkit-transform: scale(0.98);
  -ms-transform: scale(0.98);
  transform: scale(0.98);
  background-color: #0f0;
}

<button
  className="emotion-0"
>
  Button
</button>
`;
